AX: Merge AccessibilityARIATableCell into AccessibilityTableCell

hoffmanjoshua opened the pull request:
#### d7dfe92466569af3932ad2535f5a619b24f7c957
<pre>
AX: Merge AccessibilityARIATableCell into AccessibilityTableCell
<a href="https://bugs.webkit.org/show_bug.cgi?id=296627">https://bugs.webkit.org/show_bug.cgi?id=296627</a>
<a href="https://rdar.apple.com/157023868">rdar://157023868</a>

Reviewed by Tyler Wilcock.

Move the logic from AccessibilityARIATableCell into AccessibilityTableCell
to flatten our inheritance hierarchy.

* Source/WebCore/Sources.txt:
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::createObjectFromRenderer):
(WebCore::AXObjectCache::createFromNode):
* Source/WebCore/accessibility/AccessibilityARIAGridCell.cpp: Removed.
* Source/WebCore/accessibility/AccessibilityARIAGridCell.h: Removed.
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::isAccessibilityARIAGridRowInstance const):
* Source/WebCore/accessibility/AccessibilityTableCell.cpp:
(WebCore::AccessibilityTableCell::AccessibilityTableCell):
(WebCore::AccessibilityTableCell::create):
(WebCore::AccessibilityTableCell::parentTable const):
(WebCore::AccessibilityTableCell::readOnlyValue const):
* Source/WebCore/accessibility/AccessibilityTableCell.h:
(WebCore::AccessibilityTableCell::isARIAGridCell const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:

Canonical link: <a href="https://commits.webkit.org/298019@main">https://commits.webkit.org/298019@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/65d62ae031c479eee1b5bddef5b0ddd9b04daf31

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113899 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/33651 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24110 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120063 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/64682 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/115788 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34279 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42212 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86572 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/64682 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116847 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27282 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102299 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66946 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26483 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20431 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63781 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96650 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20545 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123294 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40943 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30488 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95402 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41316 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98511 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95175 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24275 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40313 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18120 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/37079 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40812 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46322 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40450 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43750 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42207 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AccessibilityTableCell.cpp:
```diff
@@ -125,6 +137,19 @@ AccessibilityTable* AccessibilityTableCell::parentTable() const
+String AccessibilityTableCell::readOnlyValue() const
+{
+    if (hasAttribute(aria_readonlyAttr))
+        return getAttribute(aria_readonlyAttr).string().convertToASCIILowercase();
```
I know you're just porting this as-is from the `AccessibilityARIAGridCell`. But `hasAttribute` and `getAttribute` do a lot of work, and it's mostly the same. Is it possible to implement this using just `getAttribute` and checking for `nullAtom()` rather than using `hasAttribute`?

hoffmanjoshua replied:
Fixed in latest commit!

twilco commented on Source/WebCore/accessibility/AccessibilityTableCell.h:
```diff
@@ -96,6 +98,7 @@ class AccessibilityTableCell : public AccessibilityRenderObject {
     // reflect how many rows the cell spans in the rendered table.
     // Default to 1, as the cell should span at least the row it starts in.
     unsigned m_effectiveRowSpan { 1 };
+    bool m_isARIAGridCell { false };
```
Did you run `dump-class-layout` before placing this `bool`?

hoffmanjoshua replied:
Moved to AccessibilityNodeObject to fit into some padding there

hoffmanjoshua committed:
AX: Merge AccessibilityARIATableCell into AccessibilityTableCell
https://bugs.webkit.org/show_bug.cgi?id=296627
rdar://157023868

Reviewed by Tyler Wilcock.

Move the logic from AccessibilityARIATableCell into AccessibilityTableCell
to flatten our inheritance hierarchy.

* Source/WebCore/Sources.txt:
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::createObjectFromRenderer):
(WebCore::AXObjectCache::createFromNode):
* Source/WebCore/accessibility/AccessibilityARIAGridCell.cpp: Removed.
* Source/WebCore/accessibility/AccessibilityARIAGridCell.h: Removed.
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::isAccessibilityARIAGridRowInstance const):
* Source/WebCore/accessibility/AccessibilityTableCell.cpp:
(WebCore::AccessibilityTableCell::AccessibilityTableCell):
(WebCore::AccessibilityTableCell::create):
(WebCore::AccessibilityTableCell::parentTable const):
(WebCore::AccessibilityTableCell::readOnlyValue const):
* Source/WebCore/accessibility/AccessibilityTableCell.h:
(WebCore::AccessibilityTableCell::isARIAGridCell const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:

Canonical link: https://commits.webkit.org/298019@main
